Replace SHARE_INSTALL by CMAKE_INSTALL_DATAROOTDIR.

berniyh opened the pull request:
CMAKE_INSTALL_DATAROOTDIR is part of GNUInstallDirs.
It defaults to share (relative to CMAKE_INSTALL_PREFIX),
but allows to be set to an absolute, thus separating
the data install dir from CMAKE_INSTALL_PREFIX, which
can be useful for cross-compiling.

Tested on Exherbo Linux, where we install data to /usr/share, but libraries and binaries are installed to /usr/$host/{bin,lib}.

houz commented:
Currently we rely on cmake 2.6 which doesn't have GNUInstallDirs yet. I would like to push that dependency so we can merge this and do other things. I don't know if we will do this before the 2.0 release or rather try to stick with what we know is working for now.

berniyh commented:
Are there still distributions that require cmake 2.6 support? Most I've seen use at least 2.8.

Apart from that. For my distribution (Exherbo Linux), I created the patches solely for Darktable 2.0. So for me, it is ok if everything older stays as is.

houz commented:
I don't know who ships what, but I don't want to find out right now, We just released rc3 and the next thing might already be the final release. Having that broken on some distribution would suck.

berniyh commented:
Sounds reasonable. If I find time I'll add a few commits further pushing the use of GNUInstallDirs.
It should simplify things here and there.

berniyh commented:
I updated the branch, the changes are outlined in the specific commits. Basically, I tried to use the directories defined by GNUInstallDirs where applicable.
I have yet to test this extensively, so please do not merge yet, I just wanted to have some feedback for now if this is going in the right direction.
I also wasn't sure about whether GNUInstallDirs will work on Apple systems, some feedback here would be appreciated.
Especially I wasn't sure about this:
if(APPLE)
  set_target_properties(darktable-cli PROPERTIES INSTALL_RPATH @loader_path/../${LIB_INSTALL}/darktable)
else(APPLE)
  set_target_properties(darktable-cli PROPERTIES INSTALL_RPATH ${CMAKE_INSTALL_FULL_LIBDIR}/darktable)
endif(APPLE)

Would the (currently) non-APPLE version work on APPLE as well?

I also had a small chat about the install paths on BSD with someone from FreeBSD.
As long as variables can be overridden, it is ok, if no system-specific handling is done in the cmake files, so I'd suggest to remove all of it, unless there is a good reason to handle it there.
GNUInstallDirs here provides the best way to extensively manipulate the paths wherever necessary and thus seems to be a good solution.

BTW, do you want to have rebased or continuously merged branches? If the former, I can rebase the branch, but obviously it would break history of local clones.

houz commented:
We should keep enough special casing to make typical compilation easy without the need to specify any variables when calling cmake (besides CMAKE_INSTALL_PREFIX).

berniyh commented:
All of the variables have default values. See here:
https://cmake.org/cmake/help/v3.0/module/GNUInstallDirs.html

Thus it should not be necessary to specify more than just the prefix. Unless you want to do some non-standard stuff, of course.
Basically nothing changed, it's just that the special cases are now handled in a standardized way by cmake.

I'm still not sure about the situation on osx, though.

LebedevRI commented on src/external/LuaAutoC/CMakeLists.txt:
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 2.6)
```
Can't change this one.

LebedevRI commented on src/external/lua/CMakeLists.txt:
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 2.6)
```
Can't change this one.

houz replied:
Even Debian/stable has 3.0.2. Where do we need 2.6?

pedrocr replied:
The current Ubuntu LTS (14.04) still has 2.8.12 so it would be nice if we could keep compatibility for that until the next LTS (16.04) is out.

McBofh replied:
On  8/01/16 10:31 AM, Pedro Côrte-Real wrote:
> In src/external/lua/CMakeLists.txt <https://github.com/darktable-org/darktable/pull/1039#discussion_r49145862>:
>
>> @@ -1,4 +1,4 @@
>> -cmake_minimum_required(VERSION 2.6)
>
> The current Ubuntu LTS (14.04) still has 2.8.12 so it would be nice if we could keep compatibility for that until the next LTS (16.04) is out.


Solaris 11.3 still has 2.8.6. Yes, not a primary development
target, but I do have users (apart from me) of this build.
I'd prefer to not have to maintain my own copy of cmake as
well as all the gnome bits :)


thanks,
James C. McPherson
--
Solaris kernel software engineer, system admin and troubleshooter
               https://www.jmcpdotcom.com/blog
Find me on LinkedIn @ http://www.linkedin.com/in/jamescmcpherson

LebedevRI replied:
I should use more verbose comments :)
What i meant is: that file is under ``src/external/`` -> probably all the changes would get overridden when updating bundled copy.
But as i now see, it looks like those 2 CMakeLists.txt are not from bundled libs, but was manually added like with rawspeed.
So all good, false alarm.

houz replied:
Ok, Solaris I understand, but an Ubuntu LTS is no reason for me in this case as we will release way after that LTS got unsupported.

pedrocr replied:
I use LTS to develop in, so if we merged this as is I'd not be able to compile master. Once 16.04 is released I'll upgrade so that's why I was suggesting we wait until that happens.

berniyh commented:
Ok, seems I was wrong and cmake 2.8 does understand GNUInstallDirs, see:
https://cmake.org/cmake/help/v2.8.12/cmake.html#module:GNUInstallDirs

Regarding the version: I can understand that Ubuntu LTS uses cmake 2.8.12 and from my point of view, it would be ok to set that as the minimum version. I'll change the tree accordingly.
2.8.6 as used on Solaris seems pretty stupid to me though. Could set this as the minimum version, but I'd rather tell them to update their stuff.
In the end, I don't think that it matters too much here, because if that cmake version is used, I'd suspect that they use older version of libraries as well and the dependency tree will break at some other point where the changes are more significant. Both for Ubuntu as well as Solaris.

LebedevRI commented:
It might be a really bad approach, but if we really need, i guess we could just import https://github.com/Kitware/CMake/blob/master/Modules/GNUInstallDirs.cmake into https://github.com/darktable-org/darktable/tree/master/cmake/modules ?

berniyh commented:
Would be possible, but it's not necessary, unless somebody comes up with a system that really needs cmake 2.6.

The change from cmake 2.8 to 3.0 was more severe than 2.6 to 2.8, so I don't think any distro/system still relies on the former.

berniyh commented:
Updated the branch (rebased on current master).

I went for cmake 2.8.5, since that is what introduced GNUInstallDirs, afaics.

parafin commented:
Works fine on OS X except one line in packaging/macosx/make-app-bundle script needs to be changed:
on line 9 change @loader_path/../lib/darktable to /usr/local/lib/darktable.
Also when changing Gentoo ebuild you must re-create the manifest, but I will need to update that stuff anyway, it's not up to date.

berniyh commented:
On 12/01/16 20:54, parafin wrote:
> Works fine on OS X except one line in packaging/macosx/make-app-bundle script
> needs to be changed:
> on line 9 change @loader_path/../lib/darktable to /usr/local/lib/darktable.
Hm, I wouldn't really like the idea of using absolute paths.

In the end it is obviously your decision, but wouldn't it be a better idea to 
use cpack to replace such a script.

> Also when changing Gentoo ebuild you must re-create the manifest, but I will
> need to update that stuff anyway, it's not up to date.
Again, it's obviously not my decision, but given that gentoo seem to provide an 
ebuild, I don't see the point in keeping that in the repository.

Anyway, I will not make changes to Gentoo (anymore) for the simple reason that 
they do not respect ownership and rebrand your changes without even asking for 
permission. So you will have to change that yourself when it is appropriate.

houz commented:
> that they do not respect ownership and rebrand your changes without even
> asking for permission.

What do you mean?

parafin commented:
Your changes introduce absolute paths in rpath, but that script depends on particular setup anyway. I'm just telling you what needs to be changed so that nothing breaks when this PR gets merged.
As for Gentoo - radhermit and me communicate when needed, there is no problem here. Some time before I created this ebuild to track git master (there was no -9999 and obviously time lag exists between updates of code and ebuild), but due to absence of time on my part it's out-of-date now and in-tree version is better.

houz commented:
Why did you make the rpath absolute? I expect that to break distribution packages.

berniyh commented:
On 03/02/16 13:57, houz wrote:
> Why did you make the rpath absolute? I expect that to break distribution packages.
>
After the last comment by parafin, I suspected it might be, so it went onto my 
todo list to check what rpath actually does and what proper handling would be there.

However, so far I didn't get to it, so it's still a todo to fix that.

houz commented:
Hint: Just leave the rpath as it was.

berniyh commented:
On 03/02/16 23:52, houz wrote:
> Hint: Just leave the rpath as it was.
Nope. LIB_INSTALL needs to die. But that was not why I hesitated to change it.

Instead, I'm wondering if the rpath thing is necessary at all. It shouldn't 
require distributions, because those rely on ld for finding the libs.
Thus, as long as ld can find the libs, everything should be fine.

Apart from that, there is no necessity for rpath to be relative, it's just a 
recommendation. I am however not convinced, that assuming $libdir (whatever it 
is called) is in the same root as $bindir is a good assumption.
It actually sounds like potential breakage because there is always some distro 
doing things different from everybody else …

So that's why I want to have a deeper look at what is going on here to find out 
if there is a better solution.

berniyh commented:
Sorry for the delay, but I had to read up some things about RPATH.
First as noted above: RPATH does not have to be relative, a full path does make sense.
I myself can only see one reason for using a relative path: a portable install version
(Which I haven't seen so far for darktable.)

As I don't know what the intention for the used RPATH settings were, I'll leave it to the project owners to decide what to do. Maybe there is something I'm not seeing here.
Other than the above, I would always recommend using absolute paths in such cases.
If you agree, I can revert the above commit.

berniyh commented:
Any more comments on this? Can it be merged for the next version?
Should I rebase it on master?

LebedevRI commented on data/CMakeLists.txt:
```diff
@@ -15,29 +15,29 @@ else ()
 #
-install(DIRECTORY "lua" DESTINATION ${SHARE_INSTALL}/darktable)
-install(FILES luarc DESTINATION ${SHARE_INSTALL}/darktable)
+install(DIRECTORY "lua" DESTINATION ${CMAKE_INSTALL_DATAROOTDIR}/darktable/)
```
What's up with sometimes adding `/` and sometimes not?
I guess for consistency it is better to have no `/` at the end.

LebedevRI commented:
@berniyh a rebase on top of current git master is definitely in order. and then please double check all the cmake code, so that all the cmakefiles is updated. 

I did not thoroughly review it, but i do like it.

Having (and using) some common ground (GNUInstallDirs) is good.
And not using it just because it bumps cmake dependency from 2.6 would be just strange, 2.6 is rather old, just like 2.8...

berniyh commented:
The / at the end wasn't intentional, so yeah I can remove that, no problem.

I'll rebase it on top of master during the next days and then check back.

berniyh commented:
Sorry this took so long, but I finally came around to rebasing this on current master. Should now merge without conflicts.

I also had a look for follow-up changes and squashed them into the respective commits.
Compiles and runs fine for me.

LebedevRI commented on src/chart/CMakeLists.txt:
```diff
@@ -15,9 +15,9 @@ set_target_properties(darktable-chart PROPERTIES LINKER_LANGUAGE C)
 if(APPLE)
-  set_target_properties(darktable-chart PROPERTIES INSTALL_RPATH @loader_path/../${LIB_INSTALL}/darktable)
+  set_target_properties(darktable-chart PROPERTIES INSTALL_RPATH @loader_path/../${CMAKE_INSTALL_LIBDIR}/darktable)
 else(APPLE)
```
The only place where INSTALL_RPATH is conditional. [1/2]

LebedevRI commented on src/external/rawspeed/CMakeLists.txt:
```diff
@@ -45,17 +45,17 @@ target_link_libraries(rawspeed ${RAWSPEED_LIBS})
   if(APPLE)
-    set_target_properties(darktable-rs-identify PROPERTIES INSTALL_RPATH @loader_path/../${LIB_INSTALL}/darktable)
+    set_target_properties(darktable-rs-identify PROPERTIES INSTALL_RPATH @loader_path/../${CMAKE_INSTALL_LIBDIR}/darktable)
   else(APPLE)
```
The only place where INSTALL_RPATH is conditional. [2/2]

LebedevRI commented:
@parafin those RPATH changes, are they ok with you, or does the RPATH needs to be apple-specific?

berniyh commented:
@LebedevRI: I think what he meant is that I removed the special APPLE handling in other cases before and didn't do it here, which is because these were added afterwards and I missed it during rebasing.
Anyway, removed it, should now be inline with the rest of the changes.

LebedevRI commented:
@berniyh otherwise looks good.
I'll wait for @parafin's confirmation and merge.

berniyh commented:
What I wasn't sure about is the WIN32 handling, there are still some hardcoded paths there, iirc.
In principle, GNUInstallDirs could be used there, but didn't really investigate.
Once it is in place, you might want to think about that.

LebedevRI commented:
@berniyh 
Don't bother :)
dt does not compile there, not without patches anyway

parafin commented:
$ORIGIN in rpath on OSX doesn't have a special meaning, it needs to be replaced with @loader_path like it was before.
So NACK in current form.

LebedevRI commented:
@berniyh from irc:
```
[00:32:39] <LebedevRI> parafin: thank you for looking at the pr
[00:33:19] <parafin> LebedevRI, np
[00:33:59] <parafin> when special handling for apple is brought back you can merge
[00:35:34] * e-anima has quit (Remote host closed the connection)
[00:36:23] <LebedevRI> $ORIGIN and @loader_path is some kind of ld variables? i wonder if $ORIGIN can be replaced with some $ORIGIN2222 which is either $ORIGIN or @loader_path
[00:36:56] <LebedevRI> s/$ORIGIN2222/${ORIGIN2222}/
[00:37:07] * Caterpillar has quit (Quit: You were not made to live as brutes, but to follow virtue and knowledge.)
[00:38:11] <parafin> yes, $ORIGIN and @loader_path are for interpretation by loader
[00:38:37] <parafin> ${ORIGIN} is a good idea
[00:38:39] <LebedevRI> i'll copy-paste that as a comment then
```

berniyh commented:
Ok. I'll do that.

Just to make sure I understood that right. The idea is to do in the root CMakeLists.txt:
if(APPLE)
  set(ORIGIN_DT @loader_path)
else(APPLE)
  set(ORIGIN_DT $ORIGIN)
endif(APPLE)

and then replace the appropriate lines in the subdir cmake files with ${ORIGIN_DT}.
(Going to use a differently named var to avoid confusion. And actually it might be necessary to use $$ORIGIN there, need to test.)

parafin commented:
That's correct, yes.

berniyh commented:
Should be good now.

Tested both the $ORIGIN and @loader_path strings and at least on my system they don't need escaping, the executables have the correct rpath stored.

LebedevRI commented:
Okay, thank you!
Seems to work fine here, let's see what more obscure distros broke :)

berniyh commented:
Feel free to CC me if some issue comes up. ;)

LebedevRI commented:
For future reference, to point packagers at: (from darix, from opensuse dt.spec)
```
 -DCMAKE_INSTALL_DATAROOTDIR="%{_datadir}" -DCMAKE_INSTALL_LIBDIR="%{_lib}" 
 and %{_lib} is the *relative* path to prefix
```
